### ISSUE TITLE
What is a target, really?

### ISSUE THREAD

**Issue by agluszak** [build-server-protocol/build-server-protocol#470]
Bazel targets map naturally to BSP targets. But for example in the sbt world it is much more complicated:
![image](https://user-images.githubusercontent.com/1674445/227979807-56cdb402-3769-4344-b8b2-a10017cd839a.png)
What if you have multiple main classes/binary artifacts/runnable targets inside a single sbt project? Shouldn't you be able to run each one separately? In cargo (Rust's default build tool, my students are implementing BSP support for it: https://github.com/cargo-bsp/cargo-bsp) there's a notion of target configurations (namely: "features") too, but we decided that - instead of each configuration of a project being a separate target - we map cargo targets to BSP targets 1-1 (and keep the configuration as a state/context of the server). So it aligns more closely with Bazel than with sbt.

 I think we should specify more clearly what a target is, so that users have better intuition about it. The Bazel/cargo definition of a binary target would be: "a binary target is the smallest unit within a project that can be run individually and produced stdout/stderr". What is a single test target would depend on the testing framework and language, but we could specify that "when a test target is run, it must produce a clear indication whether it succeeded or failed. The granularity and type (unit/integration/e2e) of the test depends on the testing framework. A test target can map for example to a single unit test or all unit tests in a given file (for example, in the Java class),"

**Comment by ckipp01** [comment]
It's an interesting question. I think some of it also comes from a bit of familiarity with the tool. When thinking of build targets, the targets that sbt produces currently actually _do make sense_ in the context of sbt. It's funny this sort of relates to a conversation that we had at the Center today about custom configurations, which can also produce targets!

If I was to greatly simplify what I think a target means, it'd be something (at least in the Scala world) that has a unique classpath. So in the case of a cross build Scala 3 / Scala 2.13 and JVM and ScalaJS project, the 8 actually make sense as each of those will have a unique classpath for both `Compile` and `Test`. There is definitely some term overloading going on in the sbt world here, so I think that's also adding some confusion.

> A test target can map for example to a single unit test or all unit tests in a given file (for example, in the Java class),"

I don't fully get this part I guess. When I think of the target I think of any test in that target that can be ran with that same classpath. Can you maybe expand on this a bit?

> What if you have multiple main classes/binary artifacts/runnable targets inside a single sbt project?

That's totally fine. A single target can have multiple main classes, but they'd again be ran/created with the same classpath since they're in the same target.

> Shouldn't you be able to run each one separately?

You can, we do this with Metals by passing in the `class` name inside of `data`.

**Comment by agluszak** [comment]
@ckipp01 so I think that projects seem to be the missing abstraction layer. See #472.

> When I think of the target I think of any test in that target that can be ran with that same classpath. Can you maybe expand on this a bit?
When *I* think of a target I think of a single class (in the JVM world).

> A single target can have multiple main classes, but they'd again be ran/created with the same classpath since they're in the same target.
Under #472 each main class would be a separate target withing a single project.

> You can, we do this with Metals by passing in the class name inside of data.
And I don't like it, because I think that's happening at a wrong level of abstraction. Again, IMO build servers should be forced to present each main class as a separate runnable target. **Otherwise build clients will need to handle this special scala-specific data kind just to be able to run different binaries!**

**Comment by ckipp01** [comment]
> @ckipp01 so I think that projects seem to be the missing abstraction layer. See https://github.com/build-server-protocol/build-server-protocol/issues/472.

I don't really think so. Take the multiple main class example. If you follow this approach you'd start creating "targets" that are completely artificial in sbt world for example and not logically really mappable to anything an sbt user would be familiar with. I think we should be careful to create new abstractions that maybe make sense for one specific tool, but would be completely artificial for others.

> When I think of a target I think of a single class (in the JVM world).

Would this cause a ton of duplication? For example if you have class A and class B wouldn't a ton of the request/response just be needless duplication?

> Again, IMO build servers should be forced to present each main class as a separate runnable target.

Again, I don't really think so, but I'm open to be convinced otherwise. Do maybe others have some thoughts?

**Comment by tgodzik** [comment]
> Again, IMO build servers should be forced to present each main class as a separate runnable target.
> Again, I don't really think so, but I'm open to be convinced otherwise. Do maybe others have some thoughts?

Some servers will not even have to support running, this is not a core thing of a build server, while compilation (building) is. And presenting each main class and also test class as a separate target will cause loads of redundant data to pollute the protocol. Additionally, would we base the logic of the BSP server here on the testing framework? That seems dangerous and way too coupled.

> When I think of a target I think of a single class (in the JVM world).

Target is not really a very unique word, so this might represent something different for everyone. There is usually a target directory for example, which will contain all classes including many main classes defined in a module.

If you feel like for Rust this new approach makes more sense then you can follow it, but for sure we will not be changing anything in the current BSP implementations, since it doesn't really add or solve anything for us.

**Comment by adpi2** [comment]
> And I don't like it, because I think that's happening at a wrong level of abstraction. Again, IMO build servers should be forced to present each main class as a separate runnable target. Otherwise build clients will need to handle this special scala-specific data kind just to be able to run different binaries!

I agree with @ckipp01 and @tgodzik that the build target is the right level of abstraction at the build tool level. I think that what you call `project` in #472 is the same thing as what we call `buildTarget`. So a build target is not a runnable target but rather a compile target: the minimal thing that a build tool can compile (or build). In sbt I cannot compile just one main class, I have to compile all the main classes of a build target at once.

But you made a point saying that we need to handle a special scala-specific data kind to be able to run different main classes. So maybe what we are missing is the notion of `runTarget`, that should be language-agnostic. So a `runTarget` in the JVM world would be a main class and it would be something different in Rust.

In short, the build would contain many `buildTarget`s, and each `buildTarget` would contain many `runTarget`s (or just one, or none).

**Comment by agluszak** [comment]
> >    When I think of a target I think of a single class (in the JVM world).

> Wouldn't this cause a ton of duplication? For example if you have class A and class B wouldn't a ton of the request/response just be needless duplication?

I meant a single *runnable* class

> we will not be changing anything in the current BSP implementations, since it doesn't really add or solve anything for us.
That seems to be a very bold statement. This discussion pertains to a new major version of the protocol anyway.

**Comment by agluszak** [comment]
> In sbt I cannot compile just one main class, I have to compile all the main classes of a build target at once.

In Bazel, for example, you can. But I agree that the notion of a "build target" has become somewhat blurred. In Bazel apparently what you (scala-tooling people) call a "build target" and what you called a "run target" in #472 mean the same thing.

> So maybe what we are missing is the notion of runTarget, that should be language-agnostic. So a runTarget in the JVM world would be a main class and it would be something different in Rust.

I agree.

> In short, the build would contain many buildTargets, and each buildTarget would contain many runTargets (or just one, or none).

That's precisely what I wanted to achieve by introducing a notion of a project :)

**Comment by agluszak** [comment]
>  I think we should be careful to create new abstractions that maybe make sense for one specific tool, but would be completely artificial for others.

Sometimes I have a feeling that all currently existing abstractions make sense only for sbt ;) I mean, it's hard to deny that BSP started as a Scala/JVM specific project and now we're trying to grow it further, which obviously causes us pain. Cargo and Bazel are way more similar to each other (despite being completely unrelated) than any of them to sbt. But at the same time I must admit that I have little knowledge about the inner workings of sbt, I'm telling only by the assumptions that are being made during creation of a BSP server for each of them.

**Comment by abrams27** [comment]
> For example, in Bazel we could treat each folder/BUILD.bazel file as a build target. And each Bazel target defined inside a build file would become a run target.

why'd we do that? the current abstractions works for bazel perfectly; on top of that bazel targets can be completely detached from their location in the project structure (potentially we can have all the targets defined in the root BUILD file)

also `run target` means that a target can be "executed" but it doesnt work for bazel libraries, or maybe im missing something here?

**Comment by abrams27** [comment]
> because in Bazel you have to define each main class/ test class as a separate Bazel (and hence BSP) target. 

well, afair it's *almost always* true, but not always - for example https://github.com/bazelbuild/rules_scala/blob/master/docs/scala_test_suite.md or just test rules with test discovery mechanizm
